lib/spack/spack/database.py

```diff
@@ -918,108 +918,123 @@ def reindex(self):
-                        " [already reconstructed from spec file]"
-                    )
-                    continue
+    def _reindex(self, old_data: Dict[str, InstallRecord]):
+        # Specs on the file system are the source of truth for record.spec. The old database values
+        # if available are the source of truth for the rest of the record.
+        assert self.layout, "Database layout must be set to reindex"
 
-                # If we arrived here it very likely means that we have external specs that are not
-                # dependencies of other specs. This may be the case for externally installed
-                # compilers or externally installed applications.
-                tty.debug(f"Reconstructing from old db: {entry.spec}")
-                try:
-                    self._add(
-                        spec=entry.spec,
-                        explicit=entry.explicit,
-                        installation_time=entry.installation_time,
-                    )
-                    processed_specs.add(entry.spec)
-                except Exception as e:
-                    # Something went wrong, so the spec was not restored from old data
-                    tty.debug(e)
+        specs_from_fs = self.layout.all_specs()
+        deprecated_for = self.layout.deprecated_for(specs_from_fs)
+
+        known_specs: List[spack.spec.Spec] = [
+            *specs_from_fs,
+            *(deprecated for _, deprecated in deprecated_for),
+            *(rec.spec for rec in old_data.values()),
+        ]
+
+        upstream_hashes = {
+            dag_hash for upstream in self.upstream_dbs for dag_hash in upstream._data
+        }
+        upstream_hashes.difference_update(spec.dag_hash() for spec in known_specs)
+
+        def create_node(edge: spack.spec.DependencySpec, is_upstream: bool):
+            if is_upstream:
+                return
+
+            self._data[edge.spec.dag_hash()] = InstallRecord(
+                spec=edge.spec.copy(deps=False),
+                path=edge.spec.external_path if edge.spec.external else None,
+                installed=edge.spec.external,
+            )
 
-            self._check_ref_counts()
+        # Store all nodes of known specs, exluding ones found in upstreams
+        tr.traverse_breadth_first_with_visitor(
+            known_specs,
+            tr.CoverNodesVisitor(
+                NoUpstreamVisitor(upstream_hashes, create_node), key=tr.by_dag_hash
+            ),
+        )
+
+        # Store the prefix and other information for specs were found on the file system
+        for s in specs_from_fs:
+            record = self._data[s.dag_hash()]
+            record.path = s.prefix
+            record.installed = True
+            record.explicit = True  # conservative assumption
+            record.installation_time = os.stat(s.prefix).st_ctime
+
+        # Deprecate specs
+        for spec, deprecated in deprecated_for:
```

Comment:
I think the generator would be consumed at this point:
```python
>>> def gen(x):
...   for ii in x:
...     yield ii
...   return x
>>> generator = gen([1, 2, 3])
>>> for v in generator:
...     print(v)
... 
1
2
3
>>> for v in generator:
...     print(v)
... 
>>> 
```

Reply:
the fact that no test fails is telling :p